contracts/DomainRegistar.sol

```diff
@@ -0,0 +1,91 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity ^0.8.24;
+
+// Uncomment this line to use console.log
+// import "hardhat/console.sol";
+
+error AccessDenied(string description);
+error NotEnoughFunds(uint8 provided, uint8 required);
+error DuplicateDomain();
+error TopLevelDomainsOnly();
+
+/**
+ * @title A simple contract for top-level domain registration
+ * @author Me
+ */
+contract DomainRegistar {
+    address public owner;
+    uint8 public domainPrice;
+    string[] private _domains;
+    mapping (address => uint16[]) private _ownerDomainsByIndex;
+
+    event DomainRegistration(address indexed _owner, address owner, string domain);
+    event PriceChange(uint8 newPrice, uint8 oldPrice);
+
+    constructor(uint8 _domainPrice) {
+        owner = msg.sender;
+        domainPrice = _domainPrice;
+    }
+
+    /**
+     * Set a new price for domain registration. Allowed for owner only
+     * @param newPrice price to be set
+     */
+    function updateDomainPrice(uint8 newPrice) public {
+        if(msg.sender != owner) {
+            revert AccessDenied("Domain price can be changed by owner only");
+        }
+        emit PriceChange(newPrice, domainPrice);
+        domainPrice = newPrice;
+    }
+
+    /**
+     * Register a new domain
+     * @param domain domain to register
+     */
+    function registerDomain(string calldata domain) public payable {
+        if(msg.value < domainPrice) {
+            revert NotEnoughFunds(uint8(msg.value), domainPrice);
+        }
+        if(!_isTopLevelDomain(domain)) {
```

Comment:
// line 50 - 55 
// this 2 validation is really expensive
// - 2 points

Reply:
cannot think of a better impl for _isTopLevelDomain except using Yul which could save some Gas, but I'm lack of knowledge for impl right now

Reply:
accepted